src/CSVMsg.chpl

```diff
@@ -0,0 +1,655 @@
+        return (subdoms, offsets, skips);
+    }
+
+    proc readTypedCSV(filenames: [] string, datasets: [?D] string, dtypes: list(string), row_counts: [] int, validFiles: [] bool, col_delim: string, st: borrowed SymTab): list((string, string, string)) throws {
+        // assumes the file has header since we were able to access type info
+        var rtnData: list((string, string, string));
+        var record_count = + reduce row_counts;
+        var (subdoms, offsets, skips) = generate_subdoms(filenames, row_counts, validFiles);
+
+        for (i, dset) in zip(D, datasets) {
+            var dtype = str2dtype(dtypes[i]);
+            select dtype {
+                when DType.Int64 {
+                    var a = makeDistArray(record_count, int);
+                    read_files_into_dist_array(a, dset, filenames, subdoms, skips, true, col_delim);
+                    var entry = new shared SymEntry(a);
+                    var rname = st.nextName();
+                    st.addEntry(rname, entry);
+                    rtnData.append((dset, "pdarray", rname));
+                }
+                when DType.UInt64 {
+                    var a = makeDistArray(record_count, uint);
+                    read_files_into_dist_array(a, dset, filenames, subdoms, skips, true, col_delim);
+                    var entry = new shared SymEntry(a);
+                    var rname = st.nextName();
+                    st.addEntry(rname, entry);
+                    rtnData.append((dset, "pdarray", rname));
+                }
+                when DType.Float64 {
+                    var a = makeDistArray(record_count, real);
+                    read_files_into_dist_array(a, dset, filenames, subdoms, skips, true, col_delim);
+                    var entry = new shared SymEntry(a);
+                    var rname = st.nextName();
+                    st.addEntry(rname, entry);
+                    rtnData.append((dset, "pdarray", rname));
+                }
+                when DType.Bool {
+                    var a = makeDistArray(record_count, bool);
+                    read_files_into_dist_array(a, dset, filenames, subdoms, skips, true, col_delim);
+                    var entry = new shared SymEntry(a);
+                    var rname = st.nextName();
+                    st.addEntry(rname, entry);
+                    rtnData.append((dset, "pdarray", rname));
+                }
+                when DType.Strings {
+                    var a = makeDistArray(record_count, string);
+                    read_files_into_dist_array(a, dset, filenames, subdoms, skips, true, col_delim);
+                    var col_lens = makeDistArray(record_count, int);
+                    forall (i, v) in zip(0..#a.size, a) {
+                        var tmp_str = v + "\x00";
+                        var vbytes = tmp_str.bytes();
+                        col_lens[i] = vbytes.size;
+                    }
+                    var offsets = (+ scan col_lens) - col_lens;
```

Comment:
it seems dangerous to have variables named `offsets` in the same proc

Reply:
Agreed. Not sure how I got away with that. Updated the 2nd instance to `str_offsets`.

Reply:
I didn't point out but we have the same thing going on in `readGenericCSV`

Reply:
updated